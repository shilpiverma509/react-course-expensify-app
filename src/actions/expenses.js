
import uuid from 'uuid';
//action generators for expenses
//ADD_EXPENSE generator
//installing uuid to generate id's now . 


export const addExpense = (
    {
        description='',
        note='',
        amount=0,
        createdAt=0
    }={} //setting default value of id as {}
)=>({
    type:'ADD_EXPENSE',
    expense:{
        id:uuid(),
        description,
        amount,
        note,
        createdAt

    }
})
//REMOVE_EXPENSE
//destructuring object
export const removeExpense = ({id}={})=>({
    type:'REMOVE_EXPENSE',
    id
})

//EDIT_EXPENSE
//id and update will not need defaults, if you don't have an id 
//and you are not updating anything, there is no need to call it
export const editExpense = (id,updates)=>({
   type:'EDIT_EXPENSE',
   id,
   updates
   
})