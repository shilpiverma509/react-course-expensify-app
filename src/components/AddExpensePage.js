import React from 'react';

const AddExpensePage =()=>{
    return(
    <div>
        This is from my add expense component
    </div>
    )};

export default AddExpensePage;    


