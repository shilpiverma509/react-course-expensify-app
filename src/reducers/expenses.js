//EXPENSES REDUCER
/*
    Reducers are responsible for managing all of states
    Reducers have two parameteres (state,action)
    State return the default state and action returns object property
    Reducer is a pure function and always return an object
*/ 
const expensesReducerDefaultState=[];
export default (state=expensesReducerDefaultState,action)=>{
    switch(action.type){
        case 'ADD_EXPENSE':
        return [
            ...state,
            action.expense
        ];
        // case 'REMOVE_EXPENSE':
        // return state.filter(({id})=>id!==action.id);
        case 'REMOVE_EXPENSE':
        return state.filter((expense)=>{
            return expense.id!==action.id
        });
        case 'EDIT_EXPENSE':
        //matching the id with action id
        return state.map((expense)=>{
            if(expense.id===action.id){
                return{
                    ...expense,
                    ...action.updates
                };
            }else{
                return expense;
            }
            
        });
        
        default:
        return state;
    }
}
