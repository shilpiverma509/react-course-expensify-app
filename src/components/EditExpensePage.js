import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense} from '../actions/expenses';
import {removeExpense} from '../actions/expenses';

//Refactor editExpensePAge to class based from classless
//Setup mapDispatchToProps editExpense and removeExpense

export class EditExpensePage extends React.Component{
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  onRemove = ()=>{
    this.props.removeExpense({id:this.props.expense.id})
    this.props.history.push('/');
    }
  render(){
    //debugger;
    return (
      <div>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button onClick={this.onRemove}>
        Remove
        </button>
    </div>
    )
  }
}

// const EditExpensePage = (props) => {
//   return (
//     <div>
//       <ExpenseForm
//         expense={props.expense}
//         onSubmit={(expense) => {
//           props.dispatch(editExpense(props.expense.id, expense));
//           props.history.push('/');
//         }}
//       />
//       <button onClick={()=>{
//         props.dispatch(removeExpense({id:props.expense.id}))
//         props.history.push('/');
//         }}>
//         Remove
//         </button>
//     </div>
//   );
//};



const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};
const mapDispatchToProps = (dispatch,props)=>({
  editExpense:(id,expense)=>dispatch(editExpense(id, expense)),
  removeExpense:(data)=>dispatch(removeExpense(data))
});


export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);
