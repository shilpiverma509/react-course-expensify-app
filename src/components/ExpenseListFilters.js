import React from 'react';
import { connect } from 'react-redux';
import {setTextFilter,sortByAmount,sortByDate,setStartDate,setEndDate} from '../actions/filters'
import { DateRangePicker } from 'react-dates'

class ExpenseListfilters extends React.Component{
    state={
        calendarFocused: null
    };
    onDatesChange = ({startDate,endDate})=>{
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch((setEndDate(endDate)));
    }
    onFocusChange = (calendarFocused)=>{
        this.setState(()=>({calendarFocused}));
    }
    render(){
        return(
        <div>
            <input type="text" 
            value={this.props.filters.text}
            onChange={(e)=>{
                this.props.dispatch(setTextFilter(e.target.value));
            }}
            />
            <select  value={this.props.filters.sortBy}
            onChange={(e)=>{
                (e.target.value==='date')?this.props.dispatch(sortByDate()):this.props.dispatch(sortByAmount());
            }}
            >
                <option value="date">Date</option>
                <option value="amount">Amount</option>
            </select>
            <DateRangePicker
                startDate={this.props.filters.startDate} 
                startDateId="start" 
                endDate={this.props.filters.endDate} 
                endDateId="end"
                onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                numberOfMonths={1}
                isOutsideRange = {()=>false}
                showClearDates={true}
                />
        </div>
        )
    }
    
}
const mapStateToProps = (state)=>{
    return{
        filters:state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListfilters);