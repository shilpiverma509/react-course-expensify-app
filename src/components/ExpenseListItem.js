import React from 'react';
import {removeExpense} from '../actions/expenses';
import {connect} from 'react-redux';


//destructuring props object
/**
 * const ExpenseListItem = (props)=>{
 * <div>
 *  {props.description}{props.amount}
 * </div>
 * }
 * props here is the individual expense array
 * remove button will dispatch an action when it will get clicked

const ExpenseListItem = (props)=>{
    <div>
        <h3>{props.description}</h3>
        <p>{props.amount} -{props.createdAt}</p>
        <button onClick={()=>{
           props.dispatch(removeExpense({id:props.id}));
        }}>Remove</button>
        
    </div>
} 
* The problem is with {props.id} 
* you can't use ES6 enhanced object literal syntax with object properties. 
* You'd need to do:{id:props.id}
*/
const ExpenseListItem = ({dispatch,id,amount,createdAt,description})=>(
    <div>
        <h3>{description}</h3>
        <p>{amount} -{createdAt}</p>
        <button onClick={()=>{
           dispatch(removeExpense({id}));
        }}>Remove</button>
        
    </div>
);
/*we don't need anything from the states.
The connect() gives ExpenseListItem access to the dispatch prop
which is all it needs.
So here the props came from the ExpenseList,then the Action modifier
was called. and that removes the expenses from the state 
*/
// const mapStatetoProps=(state)=>{
//     return{
//         expenses:state.expenses
//     }
// }

export default connect()(ExpenseListItem);