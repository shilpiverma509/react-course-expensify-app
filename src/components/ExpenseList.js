import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import SelectExpenses from '../selectors/SelectExpenses';

const ExpenseList = (props)=>
(
    <div>
        <h1>ExpenseList</h1>
        {props.expenses.map((expense)=>{
            return <ExpenseListItem key={expense.id} {...expense} />
        })}
    </div>
)
const mapStateToProps = (state)=>{
    return {
        expenses:SelectExpenses(state.expenses,state.filters)
    };
}

export default connect(mapStateToProps)(ExpenseList);

