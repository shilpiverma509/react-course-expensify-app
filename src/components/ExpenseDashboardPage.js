/**
 * goal is to list the expenses on the dashboard page
 * how do we access the store information from react componets
 */

import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
//conects your component to the redux store

const ExpenseDashboardPage =(props)=>(
    <div>
        <ExpenseListfilters />
        <ExpenseList />
    </div>
    );

export default ExpenseDashboardPage;