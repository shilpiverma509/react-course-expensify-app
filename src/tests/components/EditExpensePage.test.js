import React from 'react';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses';
import {EditExpensePage} from '../../components/EditExpensePage';


let editExpense,removeExpense,history,wrapper;

beforeEach(()=>{
    editExpense = jest.fn();
    removeExpense = jest.fn();
    //history is an object with push property on it
     history = {push:jest.fn()}
     wrapper = shallow(
        <EditExpensePage 
            editExpense={editExpense} 
            removeExpense = {removeExpense}
            history={history}
            expenses = {expenses[2]} 
        />);
});


test('should render EditExpensePage correctly',()=>{
    expect(wrapper).toMatchSnapshot();
});

test('should handle Edit Expense',()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id,expenses[2]);
    console.log(wrapper.debug());

});

test('should handle remove Expense',()=>{
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith({id:expenses[2].id});
});


