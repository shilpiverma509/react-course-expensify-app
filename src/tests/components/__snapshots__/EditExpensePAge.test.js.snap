// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<editExpensePage
  editExpense={[MockFunction]}
  expenses={
    Object {
      "amount": 4500,
      "createdAt": 345600000,
      "description": "Credit Card",
      "id": "3",
      "note": "",
    }
  }
  history={
    Object {
      "push": [MockFunction],
    }
  }
  removeExpense={[MockFunction]}
/>
`;
